### PR TITLE
Added SelectFromModel feature selection strategy

### PR DESCRIPTION
**Pull request by latture**
This PR adds sklearn's [`SelectFromModel`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectFromModel.html) feature selection strategy.

### CODE REVIEW

File: smlb/feature_selection/select_from_model_sklearn.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    TabularData,
+)
+
+
+class SelectFromModelSklearn(Features):
+    """Select features based on importance weights, scikit-learn implementation.
+
+    .. seealso::
+        See `here <https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectFromModel.html#sklearn.feature_selection.SelectFromModel>`_ for full documentation.
+
```
Comment (gregor-robinson) [inline smlb/feature_selection/select_from_model_sklearn.py:19]: I see that this file has blank lines at the end of docstrings. That's not the case in other files. It would be nice to pick one or the other project-wide, but I won't request a change here since it fits into the style of this file.

File: smlb/feature_selection/select_from_model_sklearn.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    params,
+    Data,
+    Features,
+    SupervisedLearner,
+    TabularData,
+)
+
+
+class SelectFromModelSklearn(Features):
+    """Select features based on importance weights, scikit-learn implementation.
```
Comment (gregor-robinson) [inline smlb/feature_selection/select_from_model_sklearn.py:15]:
```suggestion
    """Select features based on importance weights using the scikit-learn SelectFromModel implementation.
```

File: smlb/feature_selection/select_from_model_sklearn.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        norm_order: int = 1,
+        max_features: Optional[int] = None,
+        estimator_getter: Union[str, Callable] = "_model",
+        *args,
+        **kwargs
+    ):
+        """Initialize State.
+
+        Parameters:
+            learner: Learner that contains an estimator. The estimator must be able to provide either
```
Comment (gregor-robinson) [inline smlb/feature_selection/select_from_model_sklearn.py:36]: It'd be nice (but can wait for later) to provide the option to pass in a feature importance "getter" function that gets feature importances from the estimator. That would allow us to use other methodologies such as SHAP or permutation.

File: smlb/feature_selection/select_from_model_sklearn.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                ``feature_importances_`` or ``coef_`` attributes after fitting that are used for
+                feature selection.
+            threshold: Threshold value to use for feature selection. Features whose
+                importance is greater or equal are kept while the others are
```
Comment (gregor-robinson) [inline smlb/feature_selection/select_from_model_sklearn.py:40]:
```suggestion
                importance is greater or equal to ``threshold`` are kept while the others are
```

File: smlb/feature_selection/select_from_model_sklearn.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+        """
+
+        learner = params.instance(learner, SupervisedLearner)
+
+        is_str_or_float = lambda arg: params.any_(arg, params.string, params.real)
+        threshold = params.optional_(threshold, is_str_or_float)
+
+        prefit = params.boolean(prefit)
+        norm_order = params.integer(norm_order)
```
Comment (gregor-robinson) [inline smlb/feature_selection/select_from_model_sklearn.py:63]: Should be >= 1, right?

File: smlb/feature_selection/select_from_model_sklearn.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            threshold: Threshold value to use for feature selection. Features whose
+                importance is greater or equal are kept while the others are
+                discarded. If "median" (resp. "mean"), then the ``threshold`` value is
+                the median (resp. the mean) of the feature importances. A scaling
+                factor (e.g., "1.25*mean") may also be used. If None and if the
+                estimator has a parameter penalty set to l1, either explicitly
+                or implicitly (e.g, Lasso), the threshold used is 1e-5.
+                Otherwise, "mean" is used by default.
+            prefit: Whether a prefit model is expected to be passed into the constructor directly or not.
+            norm_order: Order of the norm used to filter the vectors of coefficients below
```
Comment (gregor-robinson) [inline smlb/feature_selection/select_from_model_sklearn.py:48]: What does this mean? I realized it's copied from sklearn, but it's really abstruse.

File: smlb/feature_selection/select_from_model_sklearn.py (hunk 1 above)
Comment (latture) [reply smlb/feature_selection/select_from_model_sklearn.py:19]: I removed them from this file. (I think I'm in the habit of adding them since they're required for our python client...)

File: smlb/feature_selection/select_from_model_sklearn.py (hunk 5 above)
Comment (latture) [reply smlb/feature_selection/select_from_model_sklearn.py:61]: Yes, I'll add `from_=1` to this check.

File: smlb/feature_selection/select_from_model_sklearn.py (hunk 6 above)
Comment (latture) [reply smlb/feature_selection/select_from_model_sklearn.py:47]: That description is not super informative. I had to dive into the source code to see what's going on. It's the order used to rescale importances before computing the threshold or comparing values to the threshold.

The overall process is 1) rescale importances via `importances = np.linalg.norm(importances, axis=0, ord=norm_order)`, 2) use the rescaled importances to compute the threshold if one wasn't explicitly set by the user and 3) remove rescaled importances below the threshold.

**Comment by latture** [comment]
Now that #33 has been merged this is ready for review.

File: smlb/feature_selection/select_from_model_sklearn.py (hunk 6 above)
Comment (latture) [reply smlb/feature_selection/select_from_model_sklearn.py:47]: I added a sentence to that effect in the parameter's docstring.

File: smlb/workflows/learning_curve_regression.py [hunk 7]
```diff
@@ -147,24 +147,22 @@ def run(self):
...
 
         # train each learner on each training set and predict validation set
         predictions = np.empty((nlearn, ntrain), dtype=PredictiveDistribution)
         for i, learner in enumerate(self._learners):
             for j, training in enumerate(training_data):
-                learner.fit(training)
-                predictions[i, j] = learner.apply(validation_data)
+                # featurize training and validation data
+                featurized_training_data = self._features.fit(training).apply(training)
+                featurized_validation_data = self._features.apply(validation_data)
```
[the remark this answers is not in the document]
Comment (gregor-robinson) [reply smlb/workflows/learning_curve_regression.py:158]: What advantage do you see in fitting to the validation set? That would make it feel a bit less like a validation to me, but maybe I don't quite understand where you're going with that.

**Review by gregor-robinson — APPROVED** [review]

File: smlb/workflows/learning_curve_regression.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (latture) [reply smlb/workflows/learning_curve_regression.py:158]: > That would make it feel a bit less like a validation to me

I had the same reaction. I don't know why it was fit to the validation data before. Maybe is was an oversight that wasn't exposed until we added featurization that changes based on the training data.
